vimeo-networking/src/main/java/com/vimeo/networking/VimeoClient.java

```diff
@@ -611,6 +617,165 @@ public void failure(VimeoError error) {
+            if (!VimeoClient.mContinuePinCodeAuthorizationRefreshCycle || now >= mExpiresInNano ||
+                authCallback == null || vimeoClient == null) {
+                if (VimeoClient.mContinuePinCodeAuthorizationRefreshCycle) {
+                    VimeoClient.mContinuePinCodeAuthorizationRefreshCycle = false;
+                    mTimer.cancel();
+                    if (authCallback != null && now >= mExpiresInNano) {
+                        VimeoError error = new VimeoError("Pin code expired.");
+                        error.setErrorCode(ErrorCode.UNABLE_TO_LOGIN_PINCODE_EXPIRED);
+                        authCallback.failure(error);
+                    }
+                }
+
+            } else {
+                Call<VimeoAccount> call =
+                        vimeoClient.mVimeoService.logInWithPinCode(vimeoClient.getBasicAuthHeader(),
+                                                                   Vimeo.DEVICE_GRANT_TYPE,
+                                                                   mPinCodeInfo.getUserCode(),
+                                                                   mPinCodeInfo.getDeviceCode(), mScope);
+                call.enqueue(new PinCodeAccountCallback(vimeoClient, authCallback, mTimer));
+            }
+        }
+    }
+
+    /**
+     * (PRIVATE: Vimeo Use Only, will not work for third-party applications)
+     * This is a method for Vimeo's internal use in association with pin code based authentication for
+     * connected but keyboardless devices such as Android TV. The method will return immediately but will run
+     * for an extended period of time asynchronously until a timeout expires, authorization has been
+     * successful, or a failure occurs. The expected application flow is that the user is initially presented
+     * a pin code (returned in pinCodeCallback), which they then enter into a special page (from
+     * a link returned in pinCodeCallback) on Vimeo.com  on a different device.  Back on the original device,
+     * the app is polling the api to check whether the pin code has been authenticated.  The `pinCodeCallback`
+     * will be called after an initial request to retrieve the pin code and activate link.
+     * This method will handle polling the api to check if the code has been activated, and it will ultimately
+     * call the completion handler (authCallback) when that happens.  If the pin code expires while we're
+     * waiting, completion handler (authCallback) will be called with an error
+     * @param pinCodeCallback {@link ModelCallback} that will receive {@link PinCodeInfo} to display
+     * @param authCallback {@link AuthCallback} that will be notified when Authorization is complete
+     * @return a call object for the pin code request
+     */
+    public Call<PinCodeInfo> logInWithPinCode(@NotNull final ModelCallback<PinCodeInfo> pinCodeCallback,
+                                              @NotNull final AuthCallback authCallback) {
```

Comment:
Be careful with AuthCallback - problems could arise if it is passed in as an anonymous inner class. If it is passed in as such it will not be held in memory since there will be no hard references to it anywhere once this method finishes returning and it will be garbage collected on the next GC cycle (which could easily happen before the network call finishes)